fpga_interchange/arch_pack_chains.cc

```diff
@@ -0,0 +1,294 @@
+}
+
+bool Arch::getClusterPlacement(ClusterId cluster, BelId root_bel,
+                         std::vector<std::pair<CellInfo *, BelId>> &placement) const
+{
+    const Context *ctx = getCtx();
+    IdString belType = getBelType(root_bel);
+
+    // Place root
+    CellInfo *root_cell = getClusterRootCell(cluster);
+    placement.push_back(std::make_pair(root_cell, root_bel));
+
+    // Place cluster
+    auto cluster_coord_cfgs = cluster_to_coord_configs.find(cluster);
+    std::vector<std::pair<ChainCoord, int>> coord_configs = cluster_coord_cfgs->second;
+    CellInfo *next_cell = root_cell;
+    NetInfo *next_net = nullptr;
+    BelId next_bel = root_bel;
+    Loc loc = getBelLocation(next_bel);
+    log_info("Cluster '%s' [%s] Placement:\n", root_cell->cluster.c_str(ctx), belType.c_str(ctx));
+    log_info("Cell '%s' placed on (%d, %d, %d)\n", next_cell->name.c_str(ctx), loc.x, loc.y, loc.z);
+    while (true) {
+        // Get pattern to find next_cell
+        auto cell_pattern = cell_pattern_map.find(next_cell->name);
+        std::pair<IdString, IdString> pattern_cfg = cell_pattern->second;
+        IdString port = pattern_cfg.second;
+        if (port == IdString()) {
+            log_error("Cluster misconfiguration! None of patterns match cell: '%s'\n", next_cell->type.c_str(ctx));
+        }
+        next_net = next_cell->ports[port].net;
+        if (next_net == nullptr || next_net->users.size() == 0) {
+            break;
+        }
+        next_cell = next_net->users.at(0).cell;
+
+        loc = getBelLocation(next_bel);
+        bool bel_found = false;
+        Loc next_loc = loc;
+        for (auto cfg : coord_configs) {
+            ChainCoord coord = cfg.first;
+            if (coord == CHAIN_X_COORD) {
+                next_loc.x += cfg.second;
+            } else {
+                next_loc.y += cfg.second;
+            }
+            // HACK / FIXME
+            if (next_loc.x > 64 || next_loc.x < 0) {
+                return false;
+            }
+            if (next_loc.y > 154 || next_loc.y < 0) {
+                return false;
+            }
+            // ------------
+            next_bel = getBelByLocation(next_loc);
+            bel_found = next_bel.index == root_bel.index;
+            if(bel_found) {
+                break;
+            }
+        }
+        if (!bel_found) {
+            // TODO: improve this message
+            log_error("Cannot place cell: '%s'\n", next_cell->name.c_str(ctx));
+        }
+
+        Loc bel_loc = getBelLocation(next_bel);
+        placement.push_back(std::make_pair(next_cell, next_bel));
+        log_info("Cell '%s' placed on (%d, %d, %d)\n", next_cell->name.c_str(ctx), bel_loc.x, bel_loc.y, bel_loc.z);
+    }
+
+    log_info("Cluster placement for root cell: '%s'\n", root_cell->name.c_str(ctx));
+
+    return true;
+}
+
+ArcBounds Arch::getClusterBounds(ClusterId cluster) const
+{
+    ArcBounds bounds(0, 0, 0, 0);
+
+    for (auto cell : sorted(getCtx()->cells)) {
```

Comment:
I think you can look up in `cluster_to_coord_configs`, you don't need to iterate over all the cells